RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2.cc

```diff
@@ -0,0 +1,196 @@
+
+  for (int i = 0; i < nCategories_; ++i) {
+      StringCutObjectSelector<reco::GsfElectron> select(categoryCutStrings[i]);
+      categoryFunctions_.push_back(select);
+  }
+
+  // Initialize GBRForests from weight files
+  init(weightFileNames);
+
+}
+
+ElectronMVAEstimatorRun2::ElectronMVAEstimatorRun2(
+        const std::string &mvaTag, const std::string &mvaName, const bool debug):
+  AnyMVAEstimatorRun2Base( edm::ParameterSet() ),
+  name_                   (mvaName),
+  tag_                    (mvaTag),
+  methodName_             ("BDTG method"),
+  debug_                  (debug) {
+  }
+
+void ElectronMVAEstimatorRun2::init(const std::vector<std::string> &weightFileNames) {
+
+  if(debug_) {
+    std::cout << " *** Inside " << name_ << tag_ << std::endl;
+  }
+
+  // Initialize GBRForests
+  if( (int)(weightFileNames.size()) != nCategories_ )
+    throw cms::Exception("MVA config failure: ")
+      << "wrong number of weightfiles" << std::endl;
+
+  gbrForests_.clear();
+  // Create a TMVA reader object for each category
+  for(int i=0; i<nCategories_; i++){
+
+    std::vector<std::string> variableNamesInCategory;
+    std::vector<int> variablesInCategory;
+
+    // Use unique_ptr so that all readers are properly cleaned up
+    // when the vector clear() is called in the destructor
+
+    gbrForests_.push_back( GBRForestTools::createGBRForest( weightFileNames[i], variableNamesInCategory ) );
+
+    nVariables_.push_back(variableNamesInCategory.size());
+
+    variables_.push_back(variablesInCategory);
+
+    if(debug_) {
+      std::cout << " *** Inside " << name_ << tag_ << std::endl;
+      std::cout << " category " << i << " with nVariables " << nVariables_[i] << std::endl;
+    }
+
+    for (int j=0; j<nVariables_[i];++j) {
+        int index = mvaVarMngr_.getVarIndex(variableNamesInCategory[j]);
+        if(index == -1) {
+            throw cms::Exception("MVA config failure: ")
+               << "Concerning " << name_ << tag_ << std::endl
+               << "Variable " << variableNamesInCategory[j]
+               << " not found in variable definition file!" << std::endl;
+        }
+        variables_[i].push_back(index);
+
+    }
+  }
+}
+
+ElectronMVAEstimatorRun2::
+~ElectronMVAEstimatorRun2(){
+}
+
+void ElectronMVAEstimatorRun2::setConsumes(edm::ConsumesCollector&& cc) const {
+  // All tokens for event content needed by this MVA
+  // Tags from the variable helper
+  for (auto &tag : mvaVarMngr_.getHelperInputTags()) {
+      cc.consumes<edm::ValueMap<float>>(tag);
+  }
+  for (auto &tag : mvaVarMngr_.getGlobalInputTags()) {
+      cc.consumes<double>(tag);
+  }
+}
+
+float ElectronMVAEstimatorRun2::
+mvaValue( const edm::Ptr<reco::Candidate>& particle, const edm::Event& iEvent) const {
+  // Try to cast the particle into a reco particle.
+  // This should work for both reco and pat.
+  const edm::Ptr<reco::GsfElectron> eleRecoPtr = ( edm::Ptr<reco::GsfElectron> )particle;
```

Comment:
This is a major problem. This does a C style cast (equivalent to a `std::reinterpret_cast`) so forces the compiler to treat the object as if it were a `edm::Ptr<reco::GsfElectron>` even when it is not. The check in the following line does not actually check the type.

What should be done is
```
auto gsfEle =dynamic_cast<reco::GsfElectron const*>(particle.get());
if( gsfEle == nullptr) {
```

Because this conversion is not proper, I saw crashes because the subsequent call to `findCategory( eleRecoPtr)` returned -1. That value was never check and then passed to `fillMVAVariables` which then used the -1 as an index lookup into an `std::vector`.